Support Propagating Changes to Multiple Resources

jGleitz opened the pull request:
This seems to work with all tests, ~except for one: `OperationRequiredRoleMappingTransformationTest#testChangeOperationRequiredRole`~

Changes:

 * `VitruviusChange`s don’t store a URI any more, but derive it from the resolved `EChange`
 * Change propagation trusts the `ChangePropagationSpecification` regarding which domain was modified
 * Copying a change is now an instance method (better OO, imo)
 * Simplify the inheritance hierarchy of `VitruviusChange`s

jGleitz commented:
corresponding application change: https://github.com/vitruv-tools/Vitruv-Applications-ComponentBasedSystems/pull/118

jGleitz commented on bundles/framework/tools.vitruv.framework.change/src/tools/vitruv/framework/change/description/impl/ConcreteChangeImpl.xtend:
```diff
@@ -1,14 +1,241 @@
+			eChange == obj.EChange
+		} 
+		else false
+	}
```
This implementation of `#equals` allows us to provide a proper `#hashCode` and nothing seems to break with it. It was based on `ECoreUtil.equals` before.

jGleitz commented:
I fixed the failing test. This was actually an error in the production code, however, I don’t know why we didn’t discover it sooner.

jGleitz commented:
I updated this branch to match the latest `master`. It is ready for review.

jGleitz commented:
This PR contains a weird race condition where models are not saved if you don’t debug through the code slowly enough.

HeikoKlare commented:
That sounds really strange. I hope that the race condition is actually "only" an issue of this PR and was not introduced by a recent one, which we did not recognize before merging it? At least, those PRs changed synchronization behavior and placement of the `save` operation for the `VirtualModel`.

jGleitz commented:
It seems to affect only the Commonalities, which could mean that the `IntermediateResourceBridgeI` broke.

jGleitz commented:
Fixed it. Somehow this error occurred when recording separately for every `ChangePropagationSpecification`

> was not introduced by a recent one, which we did not recognize before merging it?

No, it was specific to this PR. I still don’t understand why the issue sometimes did not occur when debugging, though.

HeikoKlare commented on bundles/framework/tools.vitruv.framework.change/src/tools/vitruv/framework/change/description/impl/AbstractCompositeChangeImpl.xtend:
```diff
@@ -116,32 +87,28 @@ abstract class AbstractCompositeChangeImpl<C extends VitruviusChange> extends Ab
+		if (other === this) true
+		else if (other === null) false
+		else if (other instanceof CompositeChange<?>) {
+			changes.size == other.changes.size && new HashSet(changes).containsAll(other.changes)
```
I think that the order of changes should also be matched. Actually, I think that composite changes should always be provided with an ordered set of changes rather than an arbitrary collection. Then just matching set contents would not be necessary here. Changes, by definition, can only be unordered in rare cases, because they depend on each other. For that eason, I don't think that allowing to pass a collection to a composite change provides benefits.

jGleitz replied:
I fully agree with your arguments, however, the doc of `#equals` explicitly states

> @return true, if the object is a composite change and has the same changes in any order.

That is why I kept the implementation. I am happy to change it to a `LinkedHashSet`, though.

HeikoKlare replied:
That's interesting. I guess that there was some intent to define it that way, but probably that was just some mad generalization.
If we both agree to match the order and if nothing (meanigful) breaks, I would be in favor of doing so.

jGleitz replied:
Bad news:
```java
new LinkedHashSet(Set.of(1, 2)).equals(new LinkedHashSet(Set.of(2, 1))) // true
```

jGleitz replied:
Since I don’t think that filtering out duplicate additions is important, I’ll just leave a normal list, but change `#equals` to respect the order. Is that alright?

HeikoKlare replied:
Seems like I forgot to answer to this, but you have already provided a solutions and it is fine for me.

HeikoKlare reviewed (changes requested):
Note: This review is incomplete, because I still need some time to check the `ChangePropagator` implementation. There is a bunch of changes and they are not represented as a proper diff, because the class name change is not committed as a move operation, so I will need some time to review it.

The rest of the changes are fine for me, except for using collections instead of ordered sets/lists to represent composite changes (see comment).

However, there seems to be some bug in the implementation, as the applications tests do not pass. I do not yet know which tests are affected in total, but at least the UML-Java tests fail (more precisely the ` tools.vitruv.applications.umljava.tests.uml2java.UmlToJavaClassTest.testDeleteClassImplement`).
In addition the ` tools.vitruv.applications.umljava.tests.java2uml.JavaToUmlInterfaceTest.testDeleteInterface` throws weird exceptions, but at least it does not fail.
I've not run the transitive change and Commonalities tests yet, because the build failed before.

jGleitz commented:
1. The tests failed because sometimes changes for inverse references ended up in the resulting changes of change propagation. I am not quite sure why, but I have also not fixed it.
 2. The problem I ‘fixed’ as per comment https://github.com/vitruv-tools/Vitruv/pull/412#issuecomment-777095982 was actually that the `ResourceRepository` did not create copies of the lists it hands out. I fixed that and went back to trusting `ChangePropagationSpecifications` regarding which domain was changed
 3. I first simplified `AtomicEmfChangeRecorder` because it contained quite some convoluted code.
 4. Then I removed `AtomicEmfChangeRecorder`’s functionality of removing create changes that are followed by a delete. I had multiple reasons for this change:
       1. I found the logic very hard to understand. So much so that I suspected it of having a bug, without being able to figure out whether it was really a bug or intended behaviour.
       2. I suspected the whole approach to be wrong (what happens to modifications between the create and the delete?
       2. For all the reasons I could think of that would require this logic, change recording was the wrong spot to do it. If for example, this is an optimization for change propagation, the VSUM should apply it before propagating the changes.
       3. Nothing broke when I removed the logic.

All tests pass now, as far as I can see.

jGleitz commented:
Okay, so now I have pushed two more changes:
 * make `CompositeChange`immutable. Together with the polymorphic `copy` method, this brings us one step closer to immutable changes
 * Remove `ConcreteChangeWithFixedVuri` since it is not needed for normal domains. Introduce a similar class in the Java domain instead. 

This means that this PR now also requires a change to the Java domain: https://github.com/vitruv-tools/Vitruv-Domains-ComponentBasedSystems/pull/86

All in all, this PR has gotten pretty big. I didn’t anticipate that supporting changes to multiple resources would result in so many changes to the framework. Neither was I planning to do yet another big refactoring. But hey, I think the framework improved quite a bit with this.

This should hopefully be the final state and ready for review.

jGleitz commented on bundles/framework/tools.vitruv.framework.userinteracting/src/tools/vitruv/framework/userinteraction/InternalUserInteractor.xtend:
```diff
@@ -13,6 +13,11 @@ interface InternalUserInteractor extends UserInteractor {
+	 * Removes the given {@link UserInteractionListener} from the list of listeners that get notified about interactions.
+	 */
+	def void deregisterUserInputListener(UserInteractionListener listener)
+	
```
Having this allows us to only register the current `ChangePropagation` temporarily, without having the `ChangePropagator` registered all the time. This makes it easier to avoid bugs where the `ChangePropagator` records user interactions even though it is not currently propagating.

jGleitz commented on bundles/framework/tools.vitruv.framework.change/src/tools/vitruv/framework/change/recording/NotificationRecorder.xtend:
```diff
@@ -122,34 +120,49 @@ class NotificationRecorder implements Adapter {
+				EObjectAddedEChange<?> case eChange.isContainmentInsertion:
+					removedIds -= eChange.newValueID
+			}
+		}
```
I think this part reads nicer now.

jGleitz commented on bundles/framework/tools.vitruv.framework.change/src/tools/vitruv/framework/change/recording/NotificationRecorder.xtend:
```diff
@@ -122,34 +120,49 @@ class NotificationRecorder implements Adapter {
-			val packagedDeleteChange = VitruviusChangeFactory.instance.createConcreteApplicableChange(deleteChange);
-			currentChanges.add(indexOfRemoveChange + 1, packagedDeleteChange);
-		}
+		]
```
This part got worse, unfortunately, because we now have to re-package `CompositeChange`s.

jGleitz commented on bundles/framework/tools.vitruv.framework.change/src/tools/vitruv/framework/change/recording/NotificationRecorder.xtend:
```diff
@@ -47,7 +48,7 @@ class NotificationRecorder implements Adapter {
 	override notifyChanged(Notification notification) {
 		if (notification.newValue instanceof Notifier) {
-			addToRecording(notification.newValue as Notifier);
+			recursivelyAddAdapter(notification.newValue as Notifier)
```
calling `addToRecording` here made all added objects effectively root objects.

jGleitz commented on bundles/framework/tools.vitruv.framework.change/src/tools/vitruv/framework/change/recording/AtomicEmfChangeRecorder.xtend:
```diff
@@ -1,27 +1,22 @@
+import static com.google.common.base.Preconditions.checkState
+import org.eclipse.xtend.lib.annotations.Accessors
 
 class AtomicEmfChangeRecorder {
```
This class now serves almost no purpose, except for registering objects we record at the `UuidGeneratorAndResolver`. However, I think that this might actually be the job of the views. So maybe we should remove this class altogether?

jGleitz commented:
![Bildschirmfoto vom 2021-02-12 23-21-21](https://user-images.githubusercontent.com/4305652/107828407-0febf080-6d89-11eb-803e-ef03e1c1d8c8.png) :+1:

HeikoKlare reviewed (changes requested):
Note: Applications tests (at least `transitivechange.tests`) are stil failing and I did not run the Commonalities tests yet.

jGleitz commented:
:angry:. I made sure to check all tests, but forgot about the transitive change tests. The errors are cryptic and I have no idea what’s going wrong. The Commonalities and Mappings Tests work fine, though.

jGleitz commented:
After rebasing on #419, this PR only fails the Transactional tests. A lot of those failures are related to the TUID mechanism that causes bad unloads. So I propose to wait with this PR until we migrated entirely to UUIDs.

jGleitz commented:
I merged this branch with `master` and it is now ready for review.

jGleitz commented on bundles/framework/tools.vitruv.framework.change/src/tools/vitruv/framework/change/description/impl/ConcreteChangeImpl.xtend:
```diff
@@ -1,14 +1,242 @@
+		    	'''«change.affectedFeatureString» -= «change.oldValue» (index «change.index»)'''
+		    RemoveEReference<?, ?>:
+		    	'''«change.affectedFeatureString» -= «change.oldValueString» (index «change.index»)'''
+    	}
```
For me, this new `toString` representation makes debugging Vitruv _so_ much easier.

HeikoKlare replied:
I've not tested it yet, but since the previous representation was rather unusable, I expect this to significantly improve debugging 👍

HeikoKlare commented on bundles/framework/tools.vitruv.framework.uuid/src/tools/vitruv/framework/uuid/UuidGeneratorAndResolverImpl.xtend:
```diff
@@ -187,7 +187,8 @@ class UuidGeneratorAndResolverImpl implements UuidGeneratorAndResolver {
 	override registerEObject(String uuid, EObject eObject) {
 		checkState(eObject !== null, "Object must not be null")
-		logger.debug('''Adding UUID «uuid» for EObject: «eObject»''')
+		if (logger.isDebugEnabled) logger.debug('''Adding UUID «uuid» for EObject: «eObject»''')
```
I have to admit that I don't understand what this if statement is necessary for. Is it only to avoid that the string is built even if debugging is disabled?

jGleitz replied:
Yes.

HeikoKlare replied:
Ah, yes, it became clear to me when I saw all the other places at which it was added :-)

HeikoKlare commented on bundles/framework/tools.vitruv.framework.vsum/src/tools/vitruv/framework/vsum/modelsynchronization/ChangePropagator.xtend:
```diff
@@ -0,0 +1,219 @@
+		return new ChangePropagation(this, change, changedDomain, null).propagateChanges()
+	}
+
+	def private void applySingleChange(TransactionalChange change) {
+		change.resolveBeforeAndApplyForward(uuidResolver)
+		// add all affected models to the repository
+		change.changedVURIs.forEach [resourceRepository.getModel(it)]
+	}
```
This is obsolete since #435. I think we can completely remove this method and call `change.resolveBeforeAndApplyForward(uuidResolver)` in Line 51.

jGleitz replied:
Done

HeikoKlare reviewed (commented):
Apart from the minor comments/questions, these changes are fine for me. The optimizations of changes (and their `toString` representation) as well the refactored design of the `ChangePropagator` are great improvements. I'm currently running these changes against the applications, but since you provide a patch for them as well, I expect them to work well.

jGleitz commented:
Regarding whether composite changes should take a `List` or a `Collection`: I used the most general interface possible (`Iterable` is not possible because we'll just `size` when copying). 

The collection will be copied and internally will always be a `List`. The current state allows to pass on something like a `LinkedHashSet`. That would not be possible if a `List` was required.

So I don't object to changing to `List`, but I also don't see the need.

HeikoKlare commented:
To me, it'a bit weird that one can pass an unordered collection to a change which then becomes an order assigned, which is then also used to compare the changes. If an order is important (and in my opinion it is important for changes), it should already be clarified when passing it to a composite change.
We may also pass an `Iterable` instead of a `List` to cover your issue with `LinkedHashSet`. I do not understand you point why this should not be possible. We take the `Iterable` and internally store it in a `List`, like we do now for the passed `Collection`?

HeikoKlare commented on bundles/framework/tools.vitruv.framework.change/src/tools/vitruv/framework/change/description/impl/ConcreteChangeImpl.xtend:
```diff
@@ -1,14 +1,242 @@
+		}
+	}
+
+	override getAffectedAndReferencedEObjects() {
+		switch (eChange) {
+			UpdateAttributeEChange<?>: Set.of(eChange.affectedEObject)
+			ReplaceSingleValuedEReference<?, ?>:
+				setOfNotNull(eChange.affectedEObject, eChange.oldValue, eChange.newValue)
+			InsertEReference<?, ?>: Set.of(eChange.affectedEObject, eChange.newValue)
+			RemoveEReference<?, ?>: Set.of(eChange.affectedEObject, eChange.oldValue)
+			EObjectExistenceEChange<?>: Set.of(eChange.affectedEObject)
+			InsertRootEObject<?>: Set.of(eChange.newValue)
+			RemoveRootEObject<?>: Set.of(eChange.oldValue)
+		}
+	}
```
I thought I could make a comparable found to the other `switched`, but I can't. So its probably fine to keep it as a uniform representation, especially in comparison to `getAffectedAnfReferencedEObjectIds`.

jGleitz replied:
Counter proposal: remove `EChange#getInvolvedEObjects`. The logic there is essentially the same as here and the method is (almost) not used at all.

jGleitz replied:
@HeikoKlare :point_up:  (to make sure that this is not burried)

HeikoKlare replied:
Oh, I missed your comment, I'm sorry. Yes, I agree to your proposal.

jGleitz commented:
>  I do not understand you point why this should not be possible.

It’s possible but more cumbersome. The nice thing about `List#copyOf` is that it doesn’t create a copy if the list is already read-only. But List#copyOf [only takes a `Collection`](https://docs.oracle.com/en/java/javase/15/docs/api/java.base/java/util/List.html#copyOf(java.util.Collection)).

> To me, it'a bit weird that one can pass an unordered collection to a change which then becomes an order assigned, which is then also used to compare the changes. 

> We may also pass an `Iterable` instead of a `List`

Those two statements are somewhat contradictory since `Collection` inherits from `Iterable`. So in both cases, `Collection` and `Iterable`, we’d just make a copy of the collection in the order of iteration.

jGleitz commented:
I changed the constructors to take a `List`. The factories still take an `Iterable`. Should I change this too?

jGleitz commented on bundles/framework/tools.vitruv.framework.change/src/tools/vitruv/framework/change/description/impl/CompositeTransactionalChangeImpl.xtend:
```diff
@@ -2,25 +2,20 @@ package tools.vitruv.framework.change.description.impl
-			}
-		}
-		super.removeChange(change);
+	new(Collection<? extends TransactionalChange> changes) {
```
Done

jGleitz commented on bundles/framework/tools.vitruv.framework.change/src/tools/vitruv/framework/change/description/impl/CompositeContainerChangeImpl.xtend:
```diff
@@ -2,14 +2,15 @@ package tools.vitruv.framework.change.description.impl
+import static extension edu.kit.ipd.sdq.commons.util.java.lang.IterableUtil.*
 
 class CompositeContainerChangeImpl extends AbstractCompositeChangeImpl<VitruviusChange> implements CompositeContainerChange {
+	new(Collection<? extends VitruviusChange> changes) {
```
Done

jGleitz commented on bundles/framework/tools.vitruv.framework.change/src/tools/vitruv/framework/change/description/impl/AbstractCompositeChangeImpl.xtend:
```diff
@@ -1,113 +1,79 @@
-
-	new(List<? extends C> changes) {
-		this.changes = new LinkedList<C>(changes);
+	new(Collection<? extends C> changes) {
```
Done

HeikoKlare commented:
For me, it's fine to accept an `Iterable`, so you can keep it as it is.

HeikoKlare commented on bundles/framework/tools.vitruv.framework.vsum/src/tools/vitruv/framework/vsum/modelsynchronization/ChangePropagator.xtend:
```diff
@@ -0,0 +1,213 @@
+	}
+
+	def List<PropagatedChange> propagateChange(VitruviusChange change) {
+		change.transactionalChangeSequence.forEach [resolveBeforeAndApplyForward(uuidResolver)]
```
Is this `transactionalChangeSequence` flattening still necessary at all? Since `resolveBeforeAndApplyForward` is defined for the given `VitruviusChange` is may also be fine to just call in on that change. I think this was done before the contained changed were also propagated one by one, but as far as I understand this PR simplifies the behavior to apply the complete user change and then perform change propagation for the complete change anyway, right?

jGleitz replied:
You are right with both assessments. Is it a problem that we now apply changes in a more coarse-grained manner?

Intuitively, I’d say it shouldn’t be a problem, because the initial changes are the only ones where we can make this difference. For transitive changes, all changes have already happened when propagating changes. So the new behaviour makes the behaviour of initial vs. propagated changes more consistent.

Still, I want to make sure that this doesn’t break any assumption.

HeikoKlare replied:
From my point of view, this is absolutely fine. We initially expected it to be beneficial when we propagate as fine-grained changes as possible. But this was not a central assumption but just a thing that we expected some benefit from. This became, however, less beneficial by transitive change propagation. As you mentioned, we cannot make assumptions about the change granularity when propagating them transitively anyway. So for me the modification you made is absolutely fine. In fact, the user of the `VirtualModel` can still decide about the granularity of changes to pass to the `VirtualModel`. My fear was that our transformations make implicit assumptions regarding change granularity and that many tests fail with this change, but apparantely and luckily this is not the case.

jGleitz commented:
@HeikoKlare I pushed some more cleanup:

 * Remove `EChange#getInvolvedObjects` and `ConcreteChange#affectedObjectsEquals`
 * Remove `ChangedResourceTracker`
 * Simplify applying changes in `ChangePropagator`

HeikoKlare commented:
Perfect, thanks! All changes are fine for me. I make a final run of the applications against these changes and merge them afterwards.
